### PR TITLE
WIP: SDK endpoints, little refactoring

### PR DESCRIPTION
**Pull request by lobkovilya**
Signed-off-by: Ilya Lobkov <lobkovilya@yandex.ru>

## Motivation and Context

Issue https://github.com/networkservicemesh/networkservicemesh/issues/908

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [X] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: test/integration/basic_memif_test.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// +build basic
```
Comment (nickolaev) [inline test/integration/basic_memif_test.go]: Add this in a separate patch.

Comment (nickolaev) [reply test/integration/basic_memif_test.go]: I mean PR.

File: examples/cmd/vppagent-firewall-nse/interface.go [hunk 2]
```diff
@@ -138,18 +138,21 @@ func newVppAgentXConnComposite(configuration *common.NSConfiguration) *vppAgentX
...
 }
 
 type vppAgentAclComposite struct {
-	endpoint.BaseCompositeEndpoint
+	endpoint.ChainedImpl
 	vppAgentEndpoint string
 	aclRules         map[string]string
 }
 
+func (vac *vppAgentAclComposite) GetOpaque(interface{}) interface{} {
```
Comment (nickolaev) [inline examples/cmd/vppagent-firewall-nse/interface.go]: Why not define these in ChainedImpl? DO we need to repeate this empty implemntation everywhere?

File: sdk/endpoint/interface.go [hunk 3]
```diff
@@ -18,64 +18,73 @@ package endpoint
...
-	"github.com/sirupsen/logrus"
-
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
 	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/networkservice"
 )
 
-// CompositeEndpoint is  the base service compostion interface
-type CompositeEndpoint interface {
+type NetworkService interface {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply sdk/endpoint/interface.go]: Fixed, get rid of that interface :)

File: test/integration/basic_memif_test.go (hunk 1 above)
Comment (lobkovilya) [reply test/integration/basic_memif_test.go]: Done

File: sdk/endpoint/interface.go [hunk 4]
```diff
@@ -18,64 +18,73 @@ package endpoint
...
+}
+
+func (c *ChainedImpl) GetNext() NetworkService {
+	return c.next
 }
 
-// SetSelf sets the original self struct
-func (bce *BaseCompositeEndpoint) SetSelf(self CompositeEndpoint) {
-	bce.self = self
+// BaseCompositeEndpoint is the base service compostion struct
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply sdk/endpoint/interface.go]: Fixed

File: examples/cmd/vppagent-firewall-nse/interface.go (hunk 2 above)
Comment (lobkovilya) [reply examples/cmd/vppagent-firewall-nse/interface.go]: Good idea, done

**Comment by nickolaev** [comment]
OK, merging this one.

**Pull request merged**